[Cookie Store API] Fix quick expiry issue by accounting for CFNetwork flooring of cookie creation and expiration times

RupinMittal opened the pull request:
#### 9bf64b3a9b97a77e359d6d8c70cb4f934575fbf0
<pre>
[Cookie Store API] Fix quick expiry issue by accounting for CFNetwork flooring of cookie creation and expiration times
<a href="https://bugs.webkit.org/show_bug.cgi?id=286164">https://bugs.webkit.org/show_bug.cgi?id=286164</a>
<a href="https://rdar.apple.com/143088705">rdar://143088705</a>

Reviewed by Sihui Liu.

If an event listener is registered, setting a cookie with a sub-1 second
expiry (like 300 milliseconds) should cause a change event to be fired.
Currently, it often wrongly fires as a deletion event.

The API uses WebCore::Cookies, which hold the creation and expiry times
in milliseconds. But setting the cookies with CFNetwork API means first
converting them into NSHTTPCookies, which hold these times in seconds.
CFNetwork then floors these times. This means the milliseconds are lost.
This is the cause of the problem.

Consider these two cases:

(1)
WebCore::Cookie creation time: 1024.22 milliseconds
WebCore::Cookie expiry time:   1324.22 milliseconds
Difference is 300 milliseconds --&gt; (expiry &gt; creation) This is a change event

Converting to an NSHTTPCookie means converting to seconds and then flooring:
Converting seconds gives us 1.02 and 1.32, then flooring gives us:

NSHTTPCookie creation time: 1.00 seconds
NSHTTPCookie expiry time:   1.00 seconds
Difference is 0 seconds --&gt; (expiry &lt;= creation) This is a deletion event (WRONG)

(2)
WebCore::Cookie creation time: 1924.22 milliseconds
WebCore::Cookie expiry time:   2224.22 milliseconds
Difference is 300 milliseconds --&gt; (expiry &gt; creation) This is a change event

Converting to an NSHTTPCookie means converting to seconds and then flooring:
Converting seconds gives us 1.92 and 2.22, then flooring gives us:

NSHTTPCookie creation time: 1.00 seconds
NSHTTPCookie expiry time:   2.00 seconds
Difference is 1 seconds --&gt; (expiry &gt; creation) This is a change event (CORRECT)

So if the difference between the expiry and creation is less than 1 second,
flooring may reduce the difference to 0 seconds and cause the event to incorrectly
fire as a deletion.

To fix this, if the expiry is greater than creation (expires in the future), and the
conversion to an NSHTTPCookie (conversion to seconds and flooring) will make the expiry
and creation equal, we add 1 second to the expiry time (we don&apos;t need it for expiry
times in the past since a difference of 0 is considered deletion). This way, after
flooring the difference will be 1 second. So the event will fire correctly.

Manual testing on the Cookie Store test site shows that this works as intended.

* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: <a href="https://commits.webkit.org/289245@main">https://commits.webkit.org/289245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/385486f087174f928a88a29924bc777bf4085b69

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5531 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90918 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36820 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87955 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5770 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66669 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24468 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88913 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77922 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46959 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4267 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32208 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35899 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74899 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92643 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13159 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9652 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75426 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13373 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73773 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74574 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18394 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18798 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17232 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13192 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18540 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12965 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16398 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -467,8 +467,20 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        auto expires = *options.expires;
+        if (expires > cookie.created)
+            expires += 1000.0;
+        else if (expires < cookie.created)
```
If `expires < cookie.created`, isn't the cookie deleted anyways? Why do we need this hack?

RupinMittal replied:
This is true, the cookie is deleted anyways. I put this here because even if `expires < cookie.created`, the difference could be reduced to 0 and I wasn't sure if CFNetwork would register that as a deletion. But I just tested without this and it does seem to register as a deletion. So I've removed this.

szewai commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -467,8 +467,18 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        // This can cause the onchange event to wrongly fire as a deletion instead of a change.
+        // To get around this flooring, add 1 second to the expiration if it's in the future.
+        auto expires = *options.expires;
+        if (expires > cookie.created)
```
If `created` is 1.9s and `expires` is 2.3s, with this `expires` will be 3.3s; after flooring `created` is 1s and `expires` is 3s. Is this correct?

RupinMittal replied:
Yes, that's correct.

szewai replied:
So shouldn't we add the extra 1s only if seconds are the same after flooring?
E.g. created=1.9s, expires=2.3s => created=1s, expires=2s (no increment); created=1.1s, expires=1.5s => created=1s, expires=2s

RupinMittal replied:
That's a good point. I'll make that change and re-upload.

RupinMittal replied:
Updated!

RupinMittal committed:
[Cookie Store API] Fix quick expiry issue by accounting for CFNetwork flooring of cookie creation and expiration times
https://bugs.webkit.org/show_bug.cgi?id=286164
rdar://143088705

Reviewed by Sihui Liu.

If an event listener is registered, setting a cookie with a sub-1 second
expiry (like 300 milliseconds) should cause a change event to be fired.
Currently, it often wrongly fires as a deletion event.

The API uses WebCore::Cookies, which hold the creation and expiry times
in milliseconds. But setting the cookies with CFNetwork API means first
converting them into NSHTTPCookies, which hold these times in seconds.
CFNetwork then floors these times. This means the milliseconds are lost.
This is the cause of the problem.

Consider these two cases:

(1)
WebCore::Cookie creation time: 1024.22 milliseconds
WebCore::Cookie expiry time:   1324.22 milliseconds
Difference is 300 milliseconds --> (expiry > creation) This is a change event

Converting to an NSHTTPCookie means converting to seconds and then flooring:
Converting seconds gives us 1.02 and 1.32, then flooring gives us:

NSHTTPCookie creation time: 1.00 seconds
NSHTTPCookie expiry time:   1.00 seconds
Difference is 0 seconds --> (expiry <= creation) This is a deletion event (WRONG)

(2)
WebCore::Cookie creation time: 1924.22 milliseconds
WebCore::Cookie expiry time:   2224.22 milliseconds
Difference is 300 milliseconds --> (expiry > creation) This is a change event

Converting to an NSHTTPCookie means converting to seconds and then flooring:
Converting seconds gives us 1.92 and 2.22, then flooring gives us:

NSHTTPCookie creation time: 1.00 seconds
NSHTTPCookie expiry time:   2.00 seconds
Difference is 1 seconds --> (expiry > creation) This is a change event (CORRECT)

So if the difference between the expiry and creation is less than 1 second,
flooring may reduce the difference to 0 seconds and cause the event to incorrectly
fire as a deletion.

To fix this, if the expiry is greater than creation (expires in the future), and the
conversion to an NSHTTPCookie (conversion to seconds and flooring) will make the expiry
and creation equal, we add 1 second to the expiry time (we don't need it for expiry
times in the past since a difference of 0 is considered deletion). This way, after
flooring the difference will be 1 second. So the event will fire correctly.

Manual testing on the Cookie Store test site shows that this works as intended.

* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: https://commits.webkit.org/289245@main
